clients/drcachesim/tracer/raw2trace.h

```diff
@@ -1073,33 +1112,55 @@ template <typename T> class trace_converter_t {
                 (instr->reads_memory() || instr->writes_memory()) &&
                 // No following memref for instruction-only trace type.
                 !is_instr_only_trace) {
-                for (uint j = 0; j < instr->num_mem_srcs(); j++) {
-                    error = append_memref(tls, &buf, instr, instr->mem_src_at(j), false,
-                                          reg_vals);
-                    if (!error.empty())
-                        return error;
-                    error = handle_kernel_interrupt_and_markers(
-                        tls, &buf, cur_modoffs, instr->length(), instrs_are_separate,
-                        &interrupted);
-                    if (!error.empty())
-                        return error;
-                    if (interrupted)
-                        break;
-                }
-                // We break before subsequent memrefs on an interrupt, though with
-                // today's tracer that will never happen (i#3958).
-                for (uint j = 0; !interrupted && j < instr->num_mem_dests(); j++) {
-                    error = append_memref(tls, &buf, instr, instr->mem_dest_at(j), true,
-                                          reg_vals);
-                    if (!error.empty())
-                        return error;
-                    error = handle_kernel_interrupt_and_markers(
-                        tls, &buf, cur_modoffs, instr->length(), instrs_are_separate,
-                        &interrupted);
-                    if (!error.empty())
-                        return error;
-                    if (interrupted)
-                        break;
+                if (instr->is_scatter_or_gather()) {
+                    // The instr should either load or store, but not both. Also,
+                    // it should have a single src or dest operand.
+                    DR_ASSERT(instr->num_mem_srcs() + instr->num_mem_dests() == 1);
+                    bool is_scatter = instr->num_mem_dests() == 1;
+                    bool reached_end_of_memrefs = false;
+                    // For expanded scatter/gather instrs, we do not have prior knowledge
+                    // of the number of store/load memrefs that will be present. So we
+                    // continue reading entries until we find a non-memref entry.
```

Comment:
Is it possible there's a marker in the middle, or some other non-memref, that causes a false break?  I suppose handle_kernel_interrupt_and_markers() should consume all the markers: though IIRC it bails back to the main loop in some cases.

Reply:
From my reading, there are two cases where `handle_kernel_interrupt_and_markers` bails back to the main loop.
- when there's an interrupt marker for some other module offset. This doesn't seem like it could happen in the middle of the scatter/gather memrefs, as otherwise the module offset would be the current one's (in which case, it would be processed by `handle_kernel_interrupt_and_markers`) https://github.com/DynamoRIO/dynamorio/blob/dcbaca81b9816a40fddc2b0211475a7c5a9220a3/clients/drcachesim/tracer/raw2trace.h#L1184
- or there's markers other than the interrupt marker. https://github.com/DynamoRIO/dynamorio/blob/dcbaca81b9816a40fddc2b0211475a7c5a9220a3/clients/drcachesim/tracer/raw2trace.h#L1192
I had a look at the marker list, and I don't think any of those can occur in the middle of scatter/gather memrefs: 
kernel xfer is placed only before sigret, there are many that occur only in header (timestamp, cpu Id, file type), func tracing related ones can't happen in the middle of an instr, instr count comes only in footer.

Reply:
OK.